src/anchor_in_depth/discriminator.md

```diff
@@ -0,0 +1,132 @@
+
+```
+ fn try_deserialize(buf: &mut &[u8]) -> anchor_lang::Result<Self> {
+     if buf.len() < #discriminator.len() {
+         return Err(anchor_lang::error::ErrorCode::AccountDiscriminatorNotFound.into());
+     }
+     let given_disc = &buf[..8];
+     if &#discriminator != given_disc {
+         return Err(anchor_lang::error!(anchor_lang::error::ErrorCode::AccountDiscriminatorMismatch).with_account_name(#account_name_str));
+     }
+     Self::try_deserialize_unchecked(buf)
+ }
+```
+
+Let's illustrate the importance of the discriminator with an example.
+
+Consider a program that manages two types of accounts, Account A and Account B. Both accounts are owned by the same program and have identical fields. Now, suppose you have an instruction called 'foo' that is designed to only operate on Account A.
+
+However, a user mistakenly passes Account B as an argument to the 'foo' instruction. Given that Account B shares the same owner and the same fields as Account A, how can the program detect this mistake and throw an error?
+
+This is where the discriminator comes into play. It uniquely identifies the type of an account. Even though Account A and Account B are structurally identical and share the same owner, they have different discriminators.
+
+When the 'foo' instruction gets executed, the Anchor framework checks the discriminator of the account passed as argument. If you have declared 'foo' as `foo: Account<'info, A>`, Anchor will make sure that the passed account's discriminator matches that of Account A. If the discriminators don't match (as would be the case if Account B was passed), Anchor raises an error, preventing any unintended effects on Account B.
+
+The discriminator helps Anchor to ensure that the account being processed is indeed the one expected, preventing type-related errors at runtime. This mechanism is automatically handled when you use the `Account` type in Anchor, adding an extra layer of security to your program.
+
+## Conclusion
+
+In conclusion, discriminators in Anchor play an essential role in managing and distinguishing between various types of data and account structures. They serve as unique identifiers, enabling the Anchor framework to handle data correctly during runtime. The discriminator ensures that each is treated as a distinct entity, thereby preventing any inadvertent account manipulations. This mechanism greatly enhances the robustness and security of your programs, providing reassurance that potential type-related errors are kept to a minimum.
+
+All my code you can find at anchor's github: https://github.com/coral-xyz/anchor/tree/master/lang
```

Comment:
Let's remove this line and maybe include specific permalinks to the parts of Anchor internal code that was used above.